### PR TITLE
Refactor packet data unmarshalling to use specific version

### PR DESCRIPTION
**Pull request by chatton**
## Description

closes: #6353

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/apps/transfer/ibc_module.go [hunk 1]
```diff
@@ -205,7 +203,12 @@ func (im IBCModule) OnRecvPacket(
 ) ibcexported.Acknowledgement {
 	ack := channeltypes.NewResultAcknowledgement([]byte{byte(1)})
 
-	data, ackErr := im.unmarshalPacketDataBytesToICS20V2(packet.GetData())
+	ics20Version, ok := im.keeper.GetICS4Wrapper().GetAppVersion(ctx, packet.SourcePort, packet.SourceChannel)
```
Comment (colin-axner) [inline modules/apps/transfer/ibc_module.go:206]: this should be destination port and channel (OnRecvPacket)

File: modules/apps/transfer/ibc_module.go [hunk 2]
```diff
@@ -205,7 +203,12 @@ func (im IBCModule) OnRecvPacket(
...
+	if !ok {
+		return channeltypes.NewErrorAcknowledgement(fmt.Errorf("could not retrieve app version for channel (%s, %s)", packet.SourcePort, packet.SourceChannel))
```
Comment (colin-axner) [inline modules/apps/transfer/ibc_module.go:208]: I think we should use a registered error here so the error numbers are filled in (I think this string is dropped from the ack)

**Review by colin-axner — APPROVED** [review]
ACK pending change in OnRecvPacket

Nice quick one!

File: modules/apps/transfer/ibc_module_test.go [hunk 3]
```diff
@@ -590,21 +595,52 @@ func (suite *TransferTestSuite) TestPacketDataUnmarshalerInterface() {
 			},
 			nil,
 		},
+		{
+			"failure: valid packet data multidenom without memo",
```
Comment (colin-axner) [inline modules/apps/transfer/ibc_module_test.go:599]: test case name doesn't match expected error

**Review by crodriguezvega — COMMENTED** [review]
Nice work, @chatton!

File: modules/apps/transfer/ibc_module.go [hunk 4]
```diff
@@ -174,25 +174,33 @@ func (IBCModule) OnChanCloseConfirm(
...
-	}
 
-	return types.FungibleTokenPacketDataV2{}, errorsmod.Wrapf(ibcerrors.ErrInvalidType, "cannot unmarshal ICS-20 transfer packet data")
+		if err := datav2.ValidateBasic(); err != nil {
+			return types.FungibleTokenPacketDataV2{}, err
+		}
+
+		return datav2, nil
+	default:
+		return types.FungibleTokenPacketDataV2{}, errorsmod.Wrapf(types.ErrInvalidVersion, "invalid ICS-20 version: %s", ics20Version)
```
[the remark this answers is not in the document]
Comment (chatton) [reply modules/apps/transfer/ibc_module.go:202]: ah! Good catch, will update this

File: modules/apps/transfer/ibc_module.go [hunk 5]
```diff
@@ -205,8 +213,14 @@ func (im IBCModule) OnRecvPacket(
 ) ibcexported.Acknowledgement {
 	ack := channeltypes.NewResultAcknowledgement([]byte{byte(1)})
 
-	data, ackErr := im.unmarshalPacketDataBytesToICS20V2(packet.GetData())
+	ics20Version, ok := im.keeper.GetICS4Wrapper().GetAppVersion(ctx, packet.DestinationPort, packet.DestinationChannel)
+	if !ok {
+		return channeltypes.NewErrorAcknowledgement(errorsmod.Wrapf(ibcerrors.ErrInvalidVersion, "could not retrieve app version for channel (%s, %s)", packet.SourcePort, packet.SourceChannel))
```
[the remark this answers is not in the document]
Comment (chatton) [reply modules/apps/transfer/ibc_module.go:218]: good catch! Didn't update the fields in the ack

**Pull request merged**